frontend/src/pulldasher/filtered-pulls-state.ts

```diff
@@ -27,6 +27,11 @@ export function useFilteredPullsState(pulls: Pull[]): ReturnType {
    ];
 }
 
-function filterPulls(pulls: Pull[], filters: Filters): Pull[] {
-   return Object.values(filters).reduce((pulls, filter) => pulls.filter(filter), pulls)
+function markFilteredPulls(pulls: Pull[], filters: Filters): Pull[] {
+   const filterArray = Object.values(filters);
+   const combinedFilter = (pull: Pull): boolean => filterArray.every((filter) => filter(pull));
+   return pulls.map((pull) => {
+      pull.show = combinedFilter(pull);
+      return pull;
+   });
```

Comment:
Mutating state like this is normally a no-no- 

I made this declarative by creating a new pull object with a different show property here, which eliminates the need to pass show as a separate prop when rendering a `PullCard`. I have all the code on another branch that I can push if you want to see it.

Some quick tests suggested that it takes approximately twice as long (16ms to 30ms a render), so maybe not worth the performance hit to write it that way?

Not sure how strict we want to be about state immutability; I usually prioritize it when I can.

Reply:
> Mutating state like this is normally a no-no-

Agreed. The previous pattern was more straight-forward. We just excluded filtered pulls from the arrays passed to this component (see 29e1548c8ac5cbeff4e5669aab0bc6ea2eb47587). The drag with that approach is that React then removes the entire dom structure for each card.

This state mutation is definitely a cheater approach. Hmm...   But you suggestion is a nice in-between, keeping the `pull.show` pattern, but replacing the top-level pull instance so react re-renders these components without an explicit `show` prop. Something equivalent to: `pull = {...pull, show: false}`.

There's a million ways to skin the cat and I can think of a few more, perhaps more elegant, and still performant. I'll experiment.

Thanks for the nudge!

Reply:
My pleasure! Probably not worth spending a ton of time on. Might make the code easier to maintain, but it's still pretty readable as is.

If you give me write access to the repo I'll push my branch up so you can see my implementation.